Entities/Turret.cpp

```diff
@@ -1,395 +1,84 @@
-//                  MOID register and get ID:s for itself and its children for this frame.
-
-void Turret::UpdateChildMOIDs(vector<MovableObject *> &MOIDIndex,
-                              MOID rootMOID,
-                              bool makeNewMOID)
-{
-    if (m_pMountedMO && m_pMountedMO->GetsHitByMOs())
-        m_pMountedMO->UpdateMOID(MOIDIndex, m_RootMOID, makeNewMOID);
-
-    Attachable::UpdateChildMOIDs(MOIDIndex, m_RootMOID, makeNewMOID);
-}
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Virtual method:  GetMOIDs
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Puts all MOIDs associated with this MO and all it's descendants into MOIDs vector
-// Arguments:       Vector to store MOIDs
-// Return value:    None.
-
-void Turret::GetMOIDs(std::vector<MOID> &MOIDs) const
-{
-	if (m_pMountedMO && m_pMountedMO->GetsHitByMOs())
-		m_pMountedMO->GetMOIDs(MOIDs);
-
-	Attachable::GetMOIDs(MOIDs);
-}
-
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Virtual method:  Draw
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Draws this Turret's current graphical representation to a
-//                  BITMAP of choice.
-
-void Turret::Draw(BITMAP *pTargetBitmap,
-               const Vector &targetPos,
-               DrawMode mode,
-               bool onlyPhysical) const
-{
-    if (m_pMountedMO && !m_pMountedMO->IsDrawnAfterParent())
-        m_pMountedMO->Draw(pTargetBitmap, targetPos, mode, onlyPhysical);
-
-    Attachable::Draw(pTargetBitmap, targetPos, mode, onlyPhysical);
-
-    if (m_pMountedMO && m_pMountedMO->IsDrawnAfterParent())
-        m_pMountedMO->Draw(pTargetBitmap, targetPos, mode, onlyPhysical);
-}
-
-} // namespace RTE
\ No newline at end of file
+	ConcreteClassInfo(Turret, Attachable, 20)
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void Turret::Clear() {
+		m_MountedDevice = 0;
+		m_MountedDeviceRotOffset = 0;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int Turret::Create(const Turret &reference) {
+		if (reference.m_MountedDevice) {
+			m_ReferenceHardcodedAttachableUniqueIDs.insert(reference.m_MountedDevice->GetUniqueID());
+			SetMountedDevice(dynamic_cast<Attachable *>(reference.m_MountedDevice->Clone()));
+		}
+		Attachable::Create(reference);
+
+		m_MountedDeviceRotOffset = reference.m_MountedDeviceRotOffset;
+
+		return 0;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int Turret::ReadProperty(std::string propName, Reader &reader) {
+		if (propName == "MountedDevice") {
+			RemoveAttachable(m_MountedDevice);
```

Comment:
PresetMan stuff is complicated and a pain to explain, but essentially the problem is, if I define some Turret and give it a MountedDevice that's great, no need to remove anything and the mounted device gets added as attachable. However, if I then copyof that turret and define a new MountedDevice, it needs to remove the old one first, otherwise you end up with the old one not being removed or deleted properly, putting everything in a messy state.

Reply:
Hmmmmmm.
This needs thinking in the future:tm:.